# Swimmer Assignment Problem

This notebook solves an assignment problem (problem 8.4-2, Introduction to operations research / Frederick S. Hillier, Gerald J. Lieberman.—9th ed) for a swim relay team. The objective is to assign four swimmers to four different strokes (Backstroke, Breaststroke, Butterfly, Freestyle) in a way that minimizes the total relay time, given each swimmer's time for each stroke.

The solution used the **Hungarian Algorithm** to determine the optimal assignment.

---

## Problem Description

The coach of a swim team needs to assign swimmers to a 200-yard medley relay team for the Junior Olympics. Each swimmer has recorded their best times (in seconds) for each stroke, as shown in the table below.

The times (in seconds) for each swimmer and stroke are as follows:

| Stroke      | Carl | Chris | David | Tony | Ken  |
|-------------|------|-------|-------|------|------|
| Backstroke  | 37.7 | 32.9  | 33.8  | 37.0 | 35.4 |
| Breaststroke| 43.4 | 33.1  | 42.2  | 34.7 | 41.8 |
| Butterfly   | 33.3 | 28.5  | 38.9  | 30.4 | 33.6 |
| Freestyle   | 29.2 | 26.4  | 29.6  | 28.5 | 31.1 |

Since only four swimmers are required, the goal is to assign the best combination to achieve the minimum total time for the relay team.

---

## Solution Approach

This problem was formulated as an assignment problem, where:
- **Swimmers** are the agents, and **strokes** are the tasks.
- Then an optimal way to assign four out of the five swimmers to the four strokes was determined.

---
## Explanations of the solution approach

**Formulate the Cost Matrix**
  - The Cost Matrix is the matrix where each entry represents the time a swimmer takes for a specific stroke.
  
  
**Loop over possible exclusions among the swimmers**
 - Since five swimmers are to be assigned to only four strokes,one swimmer will be excluded in each iteration by removing their column from the matrix.
 
  
**Apply the Hungarian Algorithm (using `scipy.optimize.linear_sum_assignment`)**
  - For each reduced matrix (excluding one swimmer),the Hungarian algorithm was applied to find the minimum assignment.
  - This algorithm gives us the optimal assignment for each swimmer-stroke configuration.
 
 
**Find Minimum Cost**:
  - After calculating the total time for each configuration, we then compared it to find the minimum.

In [13]:
# Import necessary libraries
import numpy as np
from scipy.optimize import linear_sum_assignment


In [14]:
# Define the cost matrix (times for each swimmer-stroke combination)
# Rows represent strokes and columns represent swimmers
cost_matrix = np.array([
    [37.7, 32.9, 33.8, 37.0, 35.4],  # Backstroke
    [43.4, 33.1, 42.2, 34.7, 41.8],  # Breaststroke
    [33.3, 28.5, 38.9, 30.4, 33.6],  # Butterfly
    [29.2, 26.4, 29.6, 28.5, 31.1],  # Freestyle
])

# Display the cost matrix
print("Cost Matrix (Swimmer Times for Each Stroke):")
print(cost_matrix)


Cost Matrix (Swimmer Times for Each Stroke):
[[37.7 32.9 33.8 37.  35.4]
 [43.4 33.1 42.2 34.7 41.8]
 [33.3 28.5 38.9 30.4 33.6]
 [29.2 26.4 29.6 28.5 31.1]]


In [16]:
# Initialize variables to store the minimum cost and best assignment configuration
min_cost = float("inf")
best_assignment = None
best_swimmer_indices = None

# Loop through each swimmer, excluding one at a time
for i in range(cost_matrix.shape[1]):
    # Create a reduced matrix by removing the i-th swimmer
    reduced_matrix = np.delete(cost_matrix, i, axis=1)
    
    # Apply the Hungarian algorithm to find the optimal assignment for the reduced matrix
    row_indices, col_indices = linear_sum_assignment(reduced_matrix)
    cost = reduced_matrix[row_indices, col_indices].sum()
    
    # Update if this configuration has a lower cost
    if cost < min_cost:
        min_cost = cost
        best_assignment = col_indices
        best_swimmer_indices = np.delete(np.arange(cost_matrix.shape[1]), i)

min_cost, best_assignment, best_swimmer_indices


(126.2, array([2, 3, 1, 0], dtype=int64), array([0, 1, 2, 3]))

In [17]:
# Display the optimal assignment results
print(f"\nMinimum Total Time: {min_cost} seconds")
print("Best Assignment (Stroke -> Swimmer):")

# Stroke and swimmer names for reference
strokes = ["Backstroke", "Breaststroke", "Butterfly", "Freestyle"]
swimmers = ["Carl", "Chris", "David", "Tony", "Ken"]

# Display each stroke's best swimmer assignment
for stroke, swimmer_index in zip(strokes, best_swimmer_indices[best_assignment]):
    print(f"{stroke}: {swimmers[swimmer_index]}")
    
# Identify the excluded swimmer
excluded_swimmer = list(set(range(5)) - set(best_swimmer_indices))[0]
print(f"\nExcluded Swimmer: {swimmers[excluded_swimmer]}")



Minimum Total Time: 126.2 seconds
Best Assignment (Stroke -> Swimmer):
Backstroke: David
Breaststroke: Tony
Butterfly: Chris
Freestyle: Carl

Excluded Swimmer: Ken


## Results

The optimal swimmer-stroke assignment achieves a minimum total time of **126.2 seconds**. This configuration is outlined below:

| Stroke      | Assigned to |
|:-------------|:-------------|
| Backstroke  | David| 
| Breaststroke| Tony| 
| Butterfly   | Chris|
| Freestyle   | Carl| 
| **Ken was excluded in the assignment**|